### PR TITLE
Introduce Go system tests

### PR DESCRIPTION
**Pull request by axw**
## Motivation/summary

Introduce new Go packages for system tests, to reduce the number of tools we use in development. The goal is to simplify writing, running, and maintaining system tests.

 - `systemtest/apmservertest`, which provides an API similar to [net/http/httptest](https://golang.org/pkg/net/http/httptest) for running apm-server
 - `systemtest`, which holds system tests. I have translated a subset of the Python-based system tests over, but the Python tests haven't yet been removed.

These packages live within a new Go Module called `github.com/elastic/apm-server/systemtest`, which provides a clear delineation between the server code and system tests. This also means that running `go test ./...` at the top level of the repo will not run the system tests. To do that, explicitly change into the `systemtest` directory and run `go test ./...` from there.

One significant difference with the Python implementation is that we now run the proper `apm-server` binary, rather than `apm-server.test` (compiled test binary). This means we can't get code coverage from the system tests, but it means that we could run the system tests with the apm-server binary that we distribute. For now we always build the `apm-server` binary, but we might change this later.

The tests continue to rely on `docker-compose` for running Elasticsearch and Kibana. This enables us to easily keep Elasticsearch and Kibana running between test runs, to avoid repeatedly waiting for them to initialise.

## Checklist

- [x] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/).
~- [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)~

I have considered changes for:
~- [ ] documentation~
~- [ ] logging (add log lines, choose appropriate log selector, etc.)~
~- [ ] metrics and monitoring (create issue for Kibana team to add metrics to visualizations, e.g. [Kibana#44001](https://github.com/elastic/kibana/issues/44001))~
~- [ ] automated tests (add tests for the code changes, all [**unit** tests](https://github.com/elastic/apm-server/blob/master/TESTING.md) pass locally)~
~- [ ] telemetry~
~- [ ] Elasticsearch Service (https://cloud.elastic.co)~
~- [ ] Elastic Cloud Enterprise (https://www.elastic.co/products/ece)~
~- [ ] Elastic Cloud on Kubernetes (https://www.elastic.co/elastic-cloud-kubernetes)~

## How to test these changes

```
cd systemtest
go test ./...
```

## Related issues

https://github.com/elastic/apm-server/issues/3318

### CODE REVIEW

**Comment by jalvz** [comment]
Yes, I also found that more often than not some test hangs, but from the logs I can't really grok why.

**Comment by axw** [comment]
Thanks for the feedback! I just pushed a fix, which I think may address the hangs you were seeing.

**Comment by axw** [comment]
> Test logs are currently only created when running the server. It would also be nice to store logs for running the server CLI, e.g. using the apikey command.

I started looking into this, and it's a little bit complicated to do it nicely. I'd defer this unless you think it's going to be absolutely necessary.

The other concern I have is that running commands should ideally not require logging, as they're short-lived operations. Important information should be surfaced in stderr by default.

> nit: it would be nice to add a helper function similar to wait_until in our python tests.

I've added request options to the ES client which should help here. In particular, you can pass `estest.WithCondition(conditionFunc)` into `Do()`. A condition function is a function with the signature `func (*esapi.Response) bool`. The request will be retried until it returns true or a timeout passes, in which case an error is returned. e.g.

```go
var result estest.SearchResult
_, err := systemtest.Elasticsearch.Search("apm-*").WithQuery(estest.TermQuery{
        Field: "processor.event",
        Value: "onboarding",
}).Do(context.Background(), &result, estest.WithCondition(result.Hits.NonEmptyCondition()))
require.NoError(t, err)
```

This will repeatedly search `apm-*` until there's at least one onboarding doc. There's a default delay of 100ms which can be overridden by passing in `estest.WithInterval`, and a default timeout of 10s which can be overridden by passing in `estest.WithTimeout`.

File: script/jenkins/linux-test.sh [hunk 1]
```diff
@@ -12,3 +12,6 @@ cleanup() {
 trap cleanup EXIT
 
 make update docker-system-tests
+
+# TODO(axw) make this part of the "system-tests" target
```
Comment (axw) [inline script/jenkins/linux-test.sh:16]: I would do this when we remove the Python system tests, removing the `docker-system-tests` target and making `system-tests` target run the Go tests in the systemtest module.

**Comment by axw** [comment]
@simitt good call, I'll do that. I was intending to leave it for a bit to build confidence. We can always revert if there's a problem though, so no need to be so conservative.

CI will already run these tests as part of the `System and Environment Tests` stage. See `./script/jenkins/linux-test.sh`.

**Comment by axw** [comment]
I've removed the old Python tests that have been reimplemented in Go. The only one that might be contentious is the Jaeger gRPC one, which does not check doc contents. I'd prefer to merge this and do that later, since we already do approval checking in beater/jaeger.

File: systemtest/apikeycmd_test.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+	assert.Equal(t, []interface{}{attrs["id"]}, result["invalidated_api_keys"])
+	assert.Equal(t, float64(0), result["error_count"])
+}
+
+func decodeJSONMap(t *testing.T, r io.Reader) map[string]interface{} {
+	var m map[string]interface{}
+	err := json.NewDecoder(r).Decode(&m)
+	require.NoError(t, err)
+	return m
+}
```
Comment (simitt) [inline systemtest/apikeycmd_test.go:114]: API Key commands are only testing the console output, but not that the API Keys are actually created, invalidated, etc. in ES.
This has been implicitly tested in the python tests by using the `wait_until` functions.

File: systemtest/apmservertest/command.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+		return "", err
+	}
+	repoRoot := filepath.Clean(strings.TrimSpace(string(output)))
+	abspath := filepath.Join(repoRoot, "apm-server")
+	if runtime.GOOS == "windows" {
+		abspath += ".exe"
+	}
+
+	log.Println("Building apm-server...")
+	cmd := exec.Command("go", "build", "-o", abspath, "./x-pack/apm-server")
```
[the remark this answers is not in the document]
Comment (axw) [reply systemtest/apmservertest/command.go:170]: Will add a TODO. We don't have any system tests for OSS today, right?

File: systemtest/apmservertest/config.go [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+	data, err := json.Marshal(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	m := make(map[string]interface{})
+	for k, v := range extra {
+		m[k] = v
+	}
+	if err := json.Unmarshal(data, &m); err != nil {
```
[the remark this answers is not in the document]
Comment (axw) [reply systemtest/apmservertest/config.go:158]: That's what I would expect too. I think there was a reason why I had it in this order, but I'm not sure why now. I'll switch the order for now, and rename to "base" if we need to switch it back.

File: systemtest/apikeycmd_test.go (hunk 2 above)
Comment (axw) [reply systemtest/apikeycmd_test.go:161]: Good catch. I've added explicit "Authenticate" requests using the generated keys. Also added code to clean up API Keys for these tests.

**Pull request merged**